File: address-validation-us-3-nodejs/getBestMatches.js
'use strict';

/**
* @module AV3AddressValidation
* @description
*   Client module for the AV3 GetBestMatchesJson operation.
*   Validates and standardizes U.S. addresses via Service Objects’ AV3 API,
*   with built?in production/trial endpoints and automatic fallback logic.
*/

import fetch from 'node-fetch';

// Base endpoints for AV3 service
const ENDPOINTS = {
    production: 'https://ws.serviceobjects.com/av3/api.svc/',
    backup: 'https://wsbackup.serviceobjects.com/av3/api.svc/',
    trial: 'https://trial.serviceobjects.com/av3/api.svc/'
};

/**
 * Build a fully?qualified URL for the GetBestMatchesJson endpoint.
 * @param {object} params       Key/value pairs of query parameters.
 * @param {string} baseUrl      Base URL (must end with '/').
 * @returns {string}            Fully?qualified URL including encoded query string.
 */
function buildGetBestMatchesUrl(params, baseUrl) {
    const url = new URL('GetBestMatchesJson', baseUrl);
    Object.entries(params).forEach(([key, value]) => {
        url.searchParams.append(key, value ?? '');
    });
    return url.toString();
}

/**
 * Perform an HTTP GET request and parse JSON response.
 * @param {string} url          URL to request.
 * @returns {Promise<any>}      Parsed JSON body.
 * @throws {Error}              On HTTP or parse errors.
 */
async function httpGetJson(url) {
    const res = await fetch(url);
    if (!res.ok) throw new Error(`HTTP ${res.status} ${res.statusText}`);
    return res.json();
}

/**
 * @typedef {object} GetBestMatchesOptions
 * @property {string} businessName  Company name to assist suite parsing.
 * @property {string} address       Primary street address (e.g., '123 Main St').
 * @property {string} [address2]    Secondary address info (e.g., 'C/O John Smith').
 * @property {string} [city]        City; required if postalCode omitted.
 * @property {string} [state]       State code or full name; required if postalCode omitted.
 * @property {string} [postalCode]  5? or 9?digit ZIP; required if city/state omitted.
 * @property {string} licenseKey    Valid AV3 license key.
 * @property {'production'|'trial'} [environment='production']
 *                                 'production' for live+backup, 'trial' for trial only.
 * @property {number} [timeoutMs=15000]
 *                                 Request timeout in milliseconds.
 */

/**
 * Retrieve standardized address candidates via AV3 GetBestMatches.
 * Automatically falls back from production to backup on error TypeCode '3'.
 * @param {GetBestMatchesOptions} opts   Operation settings.
 * @returns {Promise<object>}            Service response JSON.
 * @throws {Error}                       On unrecoverable HTTP or service errors.
 */
export async function getBestMatches(opts) {
    const {
        businessName,
        address,
        address2 = '',
        city = '',
        state = '',
        postalCode = '',
        licenseKey,
        environment = 'production',
        timeoutMs = 15000
    } = opts;

    // Assemble query parameters matching API names
    const params = {
        BusinessName: businessName,
        Address: address,
        Address2: address2,
        City: city,
        State: state,
        PostalCode: postalCode,
        LicenseKey: licenseKey
    };

    // Choose primary URL based on environment
    const baseUrl = environment === 'trial' ? ENDPOINTS.trial : ENDPOINTS.production;
    const primaryUrl = buildGetBestMatchesUrl(params, baseUrl);

    // Helper to call with timeout
    const call = url => Promise.race([
        httpGetJson(url),
        new Promise((_, rej) => setTimeout(() => rej(new Error('Timeout')), timeoutMs))
    ]);

    try {
        let result = await call(primaryUrl);
        // Fallback if production + service-level fatal error
        if (environment === 'production' && result.Error?.TypeCode === '3') {
            const backupUrl = buildGetBestMatchesUrl(params, ENDPOINTS.backup);
            result = await call(backupUrl);
        }
        return result;
    } catch (err) {
        // On network or HTTP failure, retry backup for production
        if (environment === 'production') {
            const backupUrl = buildGetBestMatchesUrl(params, ENDPOINTS.backup);
            return call(backupUrl);
        }
        throw err;
    }
}

File: address-validation-us-3-python/REST/get_best_matches.py
'''
Service Objects, Inc.

This module provides the get_best_matches function to validate and standardize US addresses
using Service Objects’ AV3 API. It handles live/trial endpoints, fallback logic, and JSON parsing.

Functions:
    get_best_matches(business_name: str,
                     address: str,
                     address_2: str,
                     city: str,
                     state: str,
                     postal_code: str,
                     license_key: str,
                     is_live: bool) -> dict:
'''  
import requests  # HTTP client for RESTful API calls

# Endpoint URLs for AV3 service
primary_url = 'https://sws.serviceobjects.com/AV3/api.svc/GetBestMatchesJson?'
backup_url = 'https://swsbackup.serviceobjects.com/AV3/api.svc/GetBestMatchesJson?'
trial_url = 'https://trial.serviceobjects.com/AV3/api.svc/GetBestMatchesJson?'


def get_best_matches(business_name: str,
                   address: str,
                   address_2: str,
                   city: str,
                   state: str,
                   postal_code: str,
                   license_key: str,
                   is_live: bool) -> dict:
    """
    Call AV3 get_best_matches API and return validation results.

    Parameters:
        business_name (str): Company name to assist suite parsing.
        address (str): Primary street address (e.g., '123 Main St'). Required.
        address_2 (str): Secondary address info (e.g., 'C/O John Smith').
        city (str): City name; required if PostalCode not provided.
        state (str): State code or full name; required if PostalCode not provided.
        postal_code (str): 5- or 9-digit ZIP; required if City/State not provided.
        license_key (str): Service Objects license key.
        is_live (bool): True for production endpoints, False for trial URL.

    Returns:
        dict: Parsed JSON response with address candidates or error info.
    """

    # Prepare query parameters for AV3 API
    params = {
        'BusinessName': business_name,
        'Address'     : address,
        'Address2'    : address_2,
        'City'        : city,
        'State'       : state,
        'PostalCode'  : postal_code,
        'LicenseKey'  : license_key
    }

    # Select the base URL: production vs trial
    url = primary_url if is_live else trial_url

    # Attempt primary (or trial) endpoint first
    try:
        response = requests.get(url, params=params, timeout=10)
        data = response.json()

        # If API returned an error in JSON payload, trigger fallback
        if 'Error' in data:
            if is_live:
                # Try backup URL when live
                response = requests.get(backup_url, params=params, timeout=10)
                data = response.json()
                # If still error, propagate exception
                if 'Error' in data:
                    raise RuntimeError(f"AV3 service error: {data['Error']}")
            else:
                # Trial mode should not fallback; error is terminal
                raise RuntimeError(f"AV3 trial error: {data['Error']}")

        # Success: return parsed JSON data
        return data

    except requests.RequestException as req_exc:
        # Network or HTTP-level error occurred
        # TODO: Instrument alerting on repeated failures
        if is_live:
            try:
                # Fallback to backup URL on network failure
                response = requests.get(backup_url, params=params, timeout=10)
                data = response.json()
                if 'Error' in data:
                    raise RuntimeError(f"AV3 backup error: {data['Error']}")
                return data
            except Exception as backup_exc:
                # Both primary and backup failed; escalate
                raise RuntimeError("AV3 service unreachable on both endpoints") from backup_exc
        else:
            # In trial mode, propagate the network exception
            raise